lib/model/model.go

```diff
@@ -468,16 +464,41 @@ func (m *model) restartFolder(from, to config.FolderConfiguration) {
 		errMsg = "restarting"
 	}
 
+	var fset *db.FileSet
+	if !to.Paused {
+		// Creating the fileset can take a long time (metadata calculation)
+		// so we do it outside of the lock.
+		fset = db.NewFileSet(to.ID, to.Filesystem(), m.db)
+	}
+
 	m.fmut.Lock()
-	m.tearDownFolderLocked(from, fmt.Errorf("%v folder %v", errMsg, to.Description()))
-	m.fmut.Unlock()
+	defer m.fmut.Unlock()
 
+	m.tearDownFolderLocked(from, fmt.Errorf("%v folder %v", errMsg, to.Description()))
 	if !to.Paused {
-		m.addFolder(to)
+		m.addFolderLocked(to, fset)
+		m.startFolderLocked(to)
 	}
 	l.Infof("%v folder %v (%v)", infoMsg, to.Description(), to.Type)
 }
 
+// Adds and starts a newly added folder (for unpaused/changed folders, user restartFolder)
+func (m *model) newFolder(cfg config.FolderConfiguration) {
+	l.Infoln("Adding new folder", cfg.Description())
+
+	// Close connections to affected devices
+	m.closeConns(cfg.DeviceIDs(), fmt.Errorf("started folder %v", cfg.Description()))
```

Comment:
Should we do the closing afterwards, given metadata calculation can take some time, and we might reconnect and start getting index updates while that calculation is happening?